python/pyspark/worker.py

```diff
@@ -208,6 +208,41 @@ def wrapped(left_key_series, left_value_series, right_key_series, right_value_se
     return lambda kl, vl, kr, vr: [(wrapped(kl, vl, kr, vr), to_arrow_type(return_type))]
 
 
+def wrap_multi_cogrouped_map_pandas_udf(f, return_type, runner_conf, argspec):
+    def wrapped(key_series, value_series):
+        import pandas as pd
+
+        dfs = [pd.concat(series, axis=1) for series in value_series]
+
+        if runner_conf.get("pass_key") == "true":
```

Comment:
I think `pass_key` is not needed here, as we have `dfs` and `argspec.args`.

When `len(argspec.args) == len(dfs)`, no key is expected, when `len(argspec.args) == len(dfs) + 1`, the key is expected.

Reply:
Yes indeed, if we don't allow users to pass vargs in UDF.

I would like to have
- capability to pass varargs as input.
- explicitness

I am waiting for more opinions in this regard. The last experimental implementation is somewhat soft blocking this PR and I would prefer to go with your last suggestion that we always have key as first arg if we can.

Reply:
With "var-args" you mean

```python
def func(*pdfs: pd.DataFrame) -> pd.DataFrame
def func_with_key(key, *pdfs: pd.DataFrame) -> pd.DataFrame
```

Then, `len(argspec.args)` will be `0`  for `func` and `1` for `func_with_key` and `argspec.args` will not be `None` in both cases. So restricting var-args cases to above signatures (not allowing `def func(pdf: pd.DataFrame, *pdfs: pd.DataFrame) -> pd.DataFrame`) should make `pass_key` redundant.

Can you give an example what you mean with "explicitness"?

Reply:
```python
def func(key, *pdfs: pd.DataFrame) -> pd.DataFrame
```
I would really like to make this as a default then we do not need any **implicit assumption** around when the first arg is key and when it is not.

If that isn't the case then I would like to keep `pass_key` as **explicit** way for user to instruct spark to treat first arg as key. Hopefully that clears up the meaning of explicitness of the API that I have in my mind. 

> Then, len(argspec.args) will be 0 for func and 1 for func_with_key and argspec.args will not be None in both cases. So restricting var-args cases to above signatures (not allowing def func(pdf: pd.DataFrame, *pdfs: pd.DataFrame) -> pd.DataFrame) should make pass_key redundant.

Seems quite complicated and not very friendly for users [higher cognitive load].